Problem Statement:
     
  Build a model which classifies patients who have breast cancer. 

 Dataset is downloaded from kaggle: https://www.kaggle.com/uciml/breast-cancer-wisconsin-data

In [29]:
#Load required libararies
import numpy as np
import pandas as pd 
import matplotlib.pyplot as plt

In [6]:
#Unzip the Breast_cancer zip file
import zipfile

with zipfile.ZipFile('breast_cancer.zip','r') as f:
    f.printdir();
    f.extractall();  #dataset is extracted into current directory

File Name                                             Modified             Size
data.csv                                       2019-09-19 21:54:02       125204


In [48]:
#load dataset
data=pd.read_csv('data.csv')
data.head(5)

id diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0    842302         M        17.99         10.38          122.80     1001.0   
1    842517         M        20.57         17.77          132.90     1326.0   
2  84300903         M        19.69         21.25          130.00     1203.0   
3  84348301         M        11.42         20.38           77.58      386.1   
4  84358402         M        20.29         14.34          135.10     1297.0   

   smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0          0.11840           0.27760          0.3001              0.14710   
1          0.08474           0.07864          0.0869              0.07017   
2          0.10960           0.15990          0.1974              0.12790   
3          0.14250           0.28390          0.2414              0.10520   
4          0.10030           0.13280          0.1980              0.10430   

   ...  texture_worst  perimeter_worst  area_worst  smoothness_worst  \
0  ...          17.33           184.60      2019.0            0.1622   
1  ...          23.41           158.80      1956.0            0.1238   
2  ...          25.53           152.50      1709.0            0.1444   
3  ...          26.50            98.87       567.7            0.2098   
4  ...          16.67           152.20      1575.0            0.1374   

   compactness_worst  concavity_worst  concave points_worst  symmetry_worst  \
0             0.6656           0.7119                0.2654          0.4601   
1             0.1866           0.2416                0.1860          0.2750   
2             0.4245           0.4504                0.2430          0.3613   
3             0.8663           0.6869                0.2575          0.6638   
4             0.2050           0.4000                0.1625          0.2364   

   fractal_dimension_worst  Unnamed: 32  
0                  0.11890          NaN  
1                  0.08902          NaN  
2                  0.08758          NaN  
3                  0.17300          NaN  
4                  0.07678          NaN  

[5 rows x 33 columns]

In [49]:
#print shape
data.shape

(569, 33)

In [50]:
#drop unsummary feature i.e Unnamed: 32
data.drop('Unnamed: 32',axis=1,inplace=True)

In [51]:
#Independent features
#drop id and diagnosis features 
inp=data.drop(['id','diagnosis'],1)
inp.head(2)

radius_mean  texture_mean  perimeter_mean  area_mean  smoothness_mean  \
0        17.99         10.38           122.8     1001.0          0.11840   
1        20.57         17.77           132.9     1326.0          0.08474   

   compactness_mean  concavity_mean  concave points_mean  symmetry_mean  \
0           0.27760          0.3001              0.14710         0.2419   
1           0.07864          0.0869              0.07017         0.1812   

   fractal_dimension_mean  ...  radius_worst  texture_worst  perimeter_worst  \
0                 0.07871  ...         25.38          17.33            184.6   
1                 0.05667  ...         24.99          23.41            158.8   

   area_worst  smoothness_worst  compactness_worst  concavity_worst  \
0      2019.0            0.1622             0.6656           0.7119   
1      1956.0            0.1238             0.1866           0.2416   

   concave points_worst  symmetry_worst  fractal_dimension_worst  
0                0.2654          0.4601                  0.11890  
1                0.1860          0.2750                  0.08902  

[2 rows x 30 columns]

In [52]:
#Dependent feature
out=data['diagnosis']
out.head(2)

0    M
1    M
Name: diagnosis, dtype: object

In [53]:
#replace M and B with 1 and 0 respectively
out=out.replace({'M':1,'B':0})

In [54]:
#Normally scaling is used for distance based algorithms(Logistic regression etc)
#But here we doing scaling bcoz for gradient descent(convergence will take place very quickly) 

from sklearn.preprocessing import StandardScaler

sc=StandardScaler()
inp_sc=sc.fit_transform(inp)
inp_sc=pd.DataFrame(inp_sc,columns=inp.columns)

inp_sc.head(2)

radius_mean  texture_mean  perimeter_mean  area_mean  smoothness_mean  \
0     1.097064     -2.073335        1.269934   0.984375         1.568466   
1     1.829821     -0.353632        1.685955   1.908708        -0.826962   

   compactness_mean  concavity_mean  concave points_mean  symmetry_mean  \
0          3.283515        2.652874             2.532475       2.217515   
1         -0.487072       -0.023846             0.548144       0.001392   

   fractal_dimension_mean  ...  radius_worst  texture_worst  perimeter_worst  \
0                2.255747  ...      1.886690      -1.359293         2.303601   
1               -0.868652  ...      1.805927      -0.369203         1.535126   

   area_worst  smoothness_worst  compactness_worst  concavity_worst  \
0    2.001237          1.307686           2.616665         2.109526   
1    1.890489         -0.375612          -0.430444        -0.146749   

   concave points_worst  symmetry_worst  fractal_dimension_worst  
0              2.296076        2.750622                 1.937015  
1              1.087084       -0.243890                 0.281190  

[2 rows x 30 columns]

In [55]:
from sklearn.model_selection import train_test_split
xtrain,xtest,ytrain,ytest=train_test_split(inp_sc,out,test_size=0.3,random_state=48)

In [56]:
#Print Shape
print("shape of xtrain :",xtrain.shape)
print("shape of xtest :",xtest.shape)
print("shape of ytrain :",ytrain.shape)
print("shape of ytest :",ytest.shape)

shape of xtrain : (398, 30)
shape of xtest : (171, 30)
shape of ytrain : (398,)
shape of ytest : (171,)


### Model Building

Typically model building in Neural network follows 2 ways 

1. Sequential model (Most preferred )
2. Functional model

In [57]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense

In [58]:
nn_classifier=Sequential()
#first hideen layer
#units=16,number of neurons in the first hidden layer
#input_dim=number of neurons in input layers
nn_classifier.add(Dense(units=16,activation='relu',input_dim=30))
#second hidden layer
#units=8 number of neurons in the second layer
#no need to mention input_dim bcoz its a sequential layer(it knows the previous layer)
nn_classifier.add(Dense(units=8,activation='relu'))
nn_classifier.add(Dense(units=1,activation='sigmoid'))

In [59]:
#Print summary
nn_classifier.summary()

Model: "sequential_5"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_14 (Dense)             (None, 16)                496       
_________________________________________________________________
dense_15 (Dense)             (None, 8)                 136       
_________________________________________________________________
dense_16 (Dense)             (None, 1)                 9         
Total params: 641
Trainable params: 641
Non-trainable params: 0
_________________________________________________________________


In [60]:
#Number of weights and bias between input and 1st hidden layers is: (30*16)+16=496
#Number of weights and bias between 1st hidden and second hidden layers is: (16*8)+8=136
##Number of weights and bias between second hidden layer and output layers is: (8*1)+1=9

In [61]:
#optimizer: which gradient descent to use(mini,stochastic,batch)    #adam is mini-batch gradient descent(it choose the learning rate smartly)
#binary_crossentropy=binary classification
#Categorical_crossentropy=multiclass classification
nn_classifier.compile(optimizer='adam',loss='binary_crossentropy',metrics=['accuracy'])

In [62]:
#batch_size: number of samples to be consider before updating weights
#epochs: number of times a full operation to run

nn_classifier.fit(xtrain,ytrain,batch_size=32,epochs=100)

Epoch 1/100
13/13 [==============================] - 0s 831us/step - loss: 0.6852 - accuracy: 0.5238
Epoch 2/100
13/13 [==============================] - 0s 1000us/step - loss: 0.6043 - accuracy: 0.7421
Epoch 3/100
13/13 [==============================] - 0s 1ms/step - loss: 0.5295 - accuracy: 0.8485
Epoch 4/100
13/13 [==============================] - 0s 1ms/step - loss: 0.4487 - accuracy: 0.9132
Epoch 5/100
13/13 [==============================] - 0s 1ms/step - loss: 0.3935 - accuracy: 0.9108
Epoch 6/100
13/13 [==============================] - 0s 1ms/step - loss: 0.3450 - accuracy: 0.9223
Epoch 7/100
13/13 [==============================] - 0s 1ms/step - loss: 0.2847 - accuracy: 0.9365
Epoch 8/100
13/13 [==============================] - 0s 1ms/step - loss: 0.2437 - accuracy: 0.9499
Epoch 9/100
13/13 [==============================] - 0s 1ms/step - loss: 0.2134 - accuracy: 0.9515
Epoch 10/100
13/13 [==============================] - 0s 1ms/step - loss: 0.1883 - accuracy: 0.9479
Epoc

13/13 [==============================] - 0s 1ms/step - loss: 0.0371 - accuracy: 0.9905
Epoch 83/100
13/13 [==============================] - 0s 1ms/step - loss: 0.0219 - accuracy: 0.9954
Epoch 84/100
13/13 [==============================] - 0s 1ms/step - loss: 0.0350 - accuracy: 0.9879
Epoch 85/100
13/13 [==============================] - 0s 1ms/step - loss: 0.0351 - accuracy: 0.9882
Epoch 86/100
13/13 [==============================] - 0s 1ms/step - loss: 0.0258 - accuracy: 0.9938
Epoch 87/100
13/13 [==============================] - 0s 1ms/step - loss: 0.0322 - accuracy: 0.9902
Epoch 88/100
13/13 [==============================] - 0s 1ms/step - loss: 0.0355 - accuracy: 0.9884
Epoch 89/100
13/13 [==============================] - 0s 1ms/step - loss: 0.0141 - accuracy: 0.9974
Epoch 90/100
13/13 [==============================] - 0s 1ms/step - loss: 0.0270 - accuracy: 0.9908
Epoch 91/100
13/13 [==============================] - 0s 1ms/step - loss: 0.0175 - accuracy: 0.9969
Epoch 92/100


In [63]:
#testing
nn_classifier.evaluate(xtest,ytest)

6/6 [==============================] - 0s 998us/step - loss: 0.0888 - accuracy: 0.9766


[0.0888427197933197, 0.9766082167625427]

In [66]:
#predictions
ypred=nn_classifier.predict(xtest).round()

*Interpretation:*

  The Model is not suffering any overfitting issue because both train and test accuracy are same. 

In [67]:
from sklearn.metrics import confusion_matrix
confusion_matrix(ytest,ypred)

array([[94,  1],
       [ 3, 73]], dtype=int64)

In [68]:
#Classification report
from sklearn.metrics import classification_report
print(classification_report(ytest,ypred))

              precision    recall  f1-score   support

           0       0.97      0.99      0.98        95
           1       0.99      0.96      0.97        76

    accuracy                           0.98       171
   macro avg       0.98      0.97      0.98       171
weighted avg       0.98      0.98      0.98       171



 *Interpretation:*
        
  The model correctly classifies patients who have malignant and benign with 98% of accuaracy.

                                                    ---thank you---